cdap-common/src/main/java/co/cask/cdap/common/conf/Configuration.java

```diff
@@ -997,6 +998,41 @@ public void setPattern(String name, Pattern pattern) {
     }
   }
 
+
+  /**
+   * Gets information about why a property was set.  Typically this is the
+   * path to the resource objects (file, URL, etc.) the property came from, but
+   * it can also indicate that it was set programatically, or because of the
+   * command line.
+   *
+   * @param name - The property name to get the source of.
+   * @return null - If the property or its source wasn't found. Otherwise,
+   * returns a list of the sources of the resource.  The older sources are
+   * the first ones in the list.  So for example if a configuration is set from
+   * the command line, and then written out to a file that is read back in the
+   * first entry would indicate that it was set from the command line, while
+   * the second one would indicate the file that the new configuration was read
+   * in from.
+   */
+  @Beta
+  public synchronized String[] getPropertySources(String name) {
```

Comment:
If it will return null, add `@Nullable`. BTW, why not return an empty array if property not found?